src/kmeans.jl

```diff
@@ -1,91 +1,115 @@
 
 const _kmeans_default_init = :kmpp
 const _kmeans_default_maxiter = 100
 const _kmeans_default_tol = 1.0e-6
 const _kmeans_default_display = :none
 
-function kmeans!(X::AbstractMatrix{T}, centers::AbstractMatrix{T};
-                 weights::Union{Nothing, AbstractVector{<:Real}}=nothing,
-                 maxiter::Integer=_kmeans_default_maxiter,
-                 tol::Real=_kmeans_default_tol,
-                 display::Symbol=_kmeans_default_display,
-                 distance::SemiMetric=SqEuclidean()) where T<:Real
-
-    m, n = size(X)
-    m2, k = size(centers)
-    m == m2 || throw(DimensionMismatch("Inconsistent array dimensions."))
+####
+#### Exported kmeans and kmeans! functions
+####
+
+function kmeans!(X::AbstractMatrix{TX},                    # in: sample matrix (p x n)
+                 centers::AbstractMatrix{TC};              # in: current centers (p x k)
+                 weights::Union{Nothing, AbstractVector{TW}}=nothing, # in: sample weights (n)
+                 maxiter::Integer=_kmeans_default_maxiter, # in: maximum number of iterations
+                 tol::Real=_kmeans_default_tol,            # in: tolerance of change at convergence
+                 display::Symbol=_kmeans_default_display,  # in: level of display
+                 distance::SemiMetric=SqEuclidean()        # in: function to compute distances
+                 ) where {TX<:Real,TC<:Real,TW<:Real}
+    p, n = size(X)
+    p2, k = size(centers)
+
+    p == p2 || throw(DimensionMismatch("Inconsistent array dimensions " *
+                                       "for `X` and `centers`."))
     (2 <= k < n) || error("k must have 2 <= k < n.")
+    if !(weights === nothing)
+        length(weights) == n || throw(DimensionMismatch("Incorrect length " *
+                                                        "of weights."))
+    end
 
     assignments = Vector{Int}(undef, n)
-    costs = Vector{T}(undef, n)
     counts = Vector{Int}(undef, k)
-    cweights = Vector{Float64}(undef, k)
 
-    _kmeans!(X, conv_weights(T, n, weights), centers,
-             assignments, costs, counts, cweights, round(Int, maxiter), tol,
+    TC2 = promote_type(TX, TC)
+    weights === nothing || (TC2 = promote_type(TC2, TW))
+    # corner case where that's still not an AbstractFloat
+    TC2 = float(TC2)
+
+    _kmeans!(X, weights, convert(Matrix{TC2}, centers),
+             assignments, counts, round(Int, maxiter), tol,
              display_level(display), distance)
 end
 
-function kmeans(X::AbstractMatrix{<:Real}, k::Int;
-                weights=nothing, init=_kmeans_default_init,
-                maxiter::Integer=_kmeans_default_maxiter,
-                tol::Real=_kmeans_default_tol,
-                display::Symbol=_kmeans_default_display,
-                distance::SemiMetric=SqEuclidean())
 
-    m, n = size(X)
+function kmeans(X::AbstractMatrix{TX}, # in: sample matrix (p x n) columns = observations
+                k::Int;                # in: number of centers
```

Comment:
```suggestion
                k::Integer;            # in: number of centers
```
Just to clarify the arguments situation. It's good to be nice to the users and relax the type requirements of API function arguments (positional and keyword) as far as possible/reasonable.
However, one have to take into account that Julia would specialize (generate a new compiled version of the method) for each new combination of arguments.
It should be fine for wrapper functions like `kmeans()`, but we should avoid unnecessary specialization on method parameters for the "actual" functions like `_kmeans!()` (e.g. `_kmeans!(..., convert(Int, maxiter), convert(K, tol))`).
Even better, we can inform Julia that we don't need to specialize on certain args of API functions (`@nospecialize(maxiter::Integer=_kmeans_default_maxiter)`) to avoid generating multiple wrappers. @nalimilan Pls correct me if I'm missing something

Reply:
I would be very happy with `centers::Any`, but, as you see, it leads to various issues :)
OTOH, `maxiter::Integer` could be just converted to `Int` without overhead or errors.
Yes, we can leave out `@nospecialize` for now.

Reply:
`@nospecialize` would mean `maxiter` would be inferred as `Any` AFAIK. That could be very bad for performance. I wouldn't worry too much about recompilation here, people are not very likely to use custom integer types and it's not the end of the world if some recompilation happens in that case (though converting to `Int` is a good idea if the wrapper already exists).

Reply:
done